Rework readme and add documentation. (#17)
Quantco/datajudge: docs/source/examples/example.rst

    import pytest
    import sqlalchemy as sa

    from datajudge import (
	Condition,
	WithinRequirement,
	BetweenRequirement
    )
    from datajudge.pytest_integration import collect_data_tests

    # We create a Requirement (within table). This object will contain
    # all the constraints we want to test on the specified table.
    # To test another table or test the same table against another table,
    # we would create another Requirement object.
    companies_req = WithinRequirement.from_table(
	db_name="example", schema_name=None, table_name="companies"
    )

   # Adding a constraint: column "name" should exist in our table
   companies_req.add_column_existence_constraint(columns=["name"])

   # Adding a constraint with a condiction
   condition = Condition(raw_string="name = 'QuantCo'")
   companies_req.add_n_rows_min_constraint(n_rows_min=1, condition=condition)

   companies_req.add_numeric_min_constraint(column="num_employees", min_value=1)

   # Creating a requirement (set of constraints) to test constraint between
   # the table "companies" and the table "companies_archive".
   companies_between_req = BetweenRequirement.from_tables(
	db_name1="example",

> 	schema_name1=None,

Author: It is currently not marked as `Optional` even though it should.

	table_name1="companies",
	db_name2="example",
	schema_name2=None,
	table_name2="companies_archive",
   )

   companies_between_req.add_row_superset_constraint(
	columns1=['name'], columns2=['name'], constant_max_missing_fraction=0
   )

> 

Reviewer: Agreed, because `pytest` is not strictly necessary. It's certainly a nice way of invoking the data tests/assertions.

> 

Author: I added a few lines but I'm not sure I have a good idea of whether that was what you were thinking of. Lmk!

Example
=======

To get started, we will create a sample database using sqlite that contains a list of companies.

.. code-block:: python

    import sqlalchemy as sa

    eng = sa.create_engine('sqlite:///example.db')

    with eng.connect() as con:
        con.execute("CREATE TABLE companies (id INTEGER PRIMARY KEY, name TEXT, num_employees INTEGER)")
        con.execute("INSERT INTO companies (name, num_employees) VALUES ('QuantCo', 100), ('Google', 150000), ('BMW', 120000), ('Apple', 145000)")
        con.execute("CREATE TABLE companies_archive (id INTEGER PRIMARY KEY, name TEXT, num_employees INTEGER)")
        con.execute("INSERT INTO companies_archive (name, num_employees) VALUES ('QuantCo', 90), ('Google', 140000), ('BMW', 110000)")

> The table "companies_archive" contains three entries:

Reviewer 2: Would move this above the SQL INSERT code. It's much easier to read the data as an actual table and the SQL code is pretty boring (most people would just skip over it)

.. list-table:: companies_archive
   :header-rows: 1

   * - id
     - name
     - num_employees
   * - 1
     - QuantCo
     - 90
   * - 2
     - Google
     - 140,000
   * - 3
     - BMW
     - 110,000

While "companies" contains an additional entry:

.. list-table:: companies
   :header-rows: 1

   * - id
     - name
     - num_employees
   * - 1
     - QuantCo
     - 100
   * - 2
     - Google
     - 150,000
   * - 3
     - BMW
     - 120,000
   * - 4
     - Apple
     - 145,000

As an example, we will run 4 tests on this table:

> - Does the table "companies" contain a column named "name"?
> - Does the table "companies" contain at least 1 entry with the name "QuantCo"?
> - Does the column "num_employees" of the "companies" table have all positive values?
> - Does the column "name" of the table "companies" contain at least all the values of
>   the corresponding column in "companies_archive"?

Reviewer 2: ```suggestion
1. Does the table "companies" contain a column named "name"?
2. Does the table "companies" contain at least 1 entry with the name "QuantCo"?
3. Does the column "num_employees" of the "companies" table have all positive values?
4. Does the column "name" of the table "companies" contain at least all the values of
   the corresponding column in "companies_archive"?
```

.. code-block:: python

    import pytest
    import sqlalchemy as sa

    from datajudge import (
        Condition,
        WithinRequirement,

>         BetweenRequirement

Reviewer 2: ```suggestion
        BetweenRequirement,
```

    )
    from datajudge.pytest_integration import collect_data_tests

    # We create a Requirement, within a table. This object will contain
    # all the constraints we want to test on the specified table.
    # To test another table or test the same table against another table,
    # we would create another Requirement object.
    companies_req = WithinRequirement.from_table(

> 	db_name="example", schema_name=None, table_name="companies"

Reviewer 2: ```suggestion
        db_name="example", schema_name=None, table_name="companies"
```

    )

>     # We add a constraint to ensure the column "name" is present in the
>     # Requirement's table.

Reviewer 2: ```suggestion
    # Constraint 1.: Does the table "companies" contain a column named "name"?
```

    companies_req.add_column_existence_constraint(columns=["name"])

>     # We add a constraint with a Condition. We ensure that at least one
>     # row satisfies the condition in the Requirement's table.

Reviewer 2: ```suggestion
    # Constraint 2.: Does the table "companies" contain at least 1 entry with the name "QuantCo"?
```

    condition = Condition(raw_string="name = 'QuantCo'")
    companies_req.add_n_rows_min_constraint(n_rows_min=1, condition=condition)

>     # We add a constraint to ensure the minimum value of the column "num_employees"
>     # is not below 1.

Reviewer 2: ```suggestion
    # Constraint 3.: Does the column "num_employees" of the "companies" table have all positive values?
```

    companies_req.add_numeric_min_constraint(column="num_employees", min_value=1)

    # We create a new Requirement, this time between different tables.
    # Concretely, we intent to test constraints between the table "companies"
    # and the table "companies_archive".

>     companies_between_req = BetweenRequirement.from_tables(
> 	 db_name1="example",
> 	 schema_name1=None,
> 	 table_name1="companies",
> 	 db_name2="example",
> 	 schema_name2=None,
> 	 table_name2="companies_archive",

Reviewer 2: ```suggestion
    companies_between_req = BetweenRequirement.from_tables(
        db_name1="example",
        schema_name1=None,
        table_name1="companies",
        db_name2="example",
        schema_name2=None,
        table_name2="companies_archive",
```

    )

>     # We add a constraint to ensure that all of the rows from the first DataSource
>     # are a subset of the second DataSource matched by the column ``name``.

Reviewer 2: ```suggestion
    # Constraint 4.: Does the column "name" of the table "companies" contain at least all the values of the corresponding column in "companies_archive"?
```
